// jasmine-galvanize-poster/index.js
var fs = require('fs');
var request = require('request');

var JSR = {}

JSR.jasmineStarted = function(suiteInfo) {
  this.numberOfSuites = suiteInfo.totalSpecsDefined;
  this.authToken = JSON.parse(fs.readFileSync(process.env.HOME + '/.gschool.json', 'utf8')).auth_token;
  this.endpoint = JSON.parse(fs.readFileSync(process.env.HOME + '/.gschool.json', 'utf8')).url;
  this.directoryName = process.env.PWD.split('/');
  this.directoryName = this.directoryName[this.directoryName.length -1];
  this.failing = 0;
  this.passing = 0;
}

JSR.suiteStarted = function (suite) {
}

JSR.suiteDone = function (suite) {
  var that = this;
  if (this.numberOfSuites == this._resultsCache.length) {
    this._resultsCache.forEach(function(data, num) {
      var testResults = JSON.parse(data);

      testResults.status === "pending" || testResults.status === "disabled" ? that.failing += 1 : undefined;

      testResults.failedExpectations.forEach(function(test){
        that.failing += 1;
        console.log("  Message:");
        console.log('\x1b[31m', '    ' + test.message ,'\x1b[0m');
        console.log("  Stack:")
        console.log(test.stack + "\n");
      });

      testResults.passedExpectations.forEach(function(data){
        that.passing +=1;
      });
    });

    var specData = {auth_token: this.authToken, directory_name: this.directoryName, passing: this.passing, failing: this.failing};

    if (this._resultsCache.length == this.numberOfSuites) {
      request.post({
        method: "POST",
        url: this.endpoint + "/student_challenges",
        type: "JSON",
        form: specData
      }).on("data", function(response) {
      }).on("error", function(response) {
      });
      console.log('\n');
      console.log(this.numberOfSuites + " specs, " + this.failing + " failing/pending");
      send();
    }
  }
}

JSR.specStarted = function (spec) {

}

JSR.specDone = function (spec) {
  this._resultsCache.push(JSON.stringify(spec, null, 2))
}

JSR.jasmineDone = function () {
}

JSR._resultsCache = []

module.exports = JSR
